KAS-3419 Link existing publication (not via MR) to agendaitem document afterwards

erikap opened the pull request:
Add support to link a publication 'not via MR' to a document on the agenda afterwards.

Search box for publications now also returns 'publications not via MR'. In case a publication 'not via MR' is selected, the publication gets relinked to the case/agenda-item-treatment the document belongs to. The dummy data (case/agenda-item-treatment) [created before](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/development/app/services/publication-service.js#L70-L79) is removed.

Also updated to condition to classify a publication as 'via MR'/'not via MR'. Instead of checking for the presence of a subcase, the presence of an agendaitem is now checked, which is more correct semantically.

erikap committed:
Add missing store service injections on routes

erikap committed:
Ensure correct treatment is linked to publication-flow when linking document

erikap committed:
Update condition to classify publication-flow as 'via MR'

Distinction between both cases can be made based on the presence of an
agendaitem that is related to the publication-flow. For publications
not via the council of ministers a dummy agenda-item-treatment is
available but it's not linked to an agendaitem in Kaleidos.

erikap committed:
Add support to relink publication-flow not via MR to an existing case

erikap committed:
Fix filter on agendaitem existence in publication-flow-selector

erikap committed:
Merge branch 'development' into feature/KAS-3419-link-publication-to-document-afterwards

erikap committed:
Update publication table columns configuration

No need to include case.subcases anymore since source is determined
based on presence of agendaitem

erikap commented:
Failing tests seem related to this PR. Tried to fix but didn't succeed yet :worried: 

I assume the problem is the search operation in [publications/publication-flow-selector.js](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1370/files#diff-5ade19de290b9b9fc2e4841596fe49cb0f3baf6da88d8f2245f7ae94c25b298f) being split into 2 queries instead of one. Tried to update test flow in [cypress/integration/unit/publication-via-MR.spec.js](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/feature/KAS-3419-link-publication-to-document-afterwards/cypress/integration/unit/publication-via-MR.spec.js#L222-L225) to

```javascript
    cy.intercept({
      method: 'GET',
      url: '/publication-flows',
      query: {
        'filter[case][:id:]': '*',
        'filter[identification][id-name]': publicationNumber2,
      },
    }).as('getFilteredByCaseAndId');
    cy.intercept({
      method: 'GET',
      url: '/publication-flows',
      query: {
        'filter[agenda-item-treatment][:has-no:agendaitem]': 'yes',
        'filter[identification][id-name]': publicationNumber2,
      },
    }).as('getFilteredNotViaMRAndId');
    cy.intercept('PATCH', '/pieces/**').as('patchPieces');
    cy.get(dependency.emberPowerSelect.searchInput).type(publicationNumber2)
      .wait(['@getFilteredByCaseAndId', '@getFilteredNotViaMRAndId']);
```

but the `wait(['@getFilteredByCaseAndId', '@getFilteredNotViaMRAndId'])` never matches. Can you help @ValenberghsSven ?

ValenberghsSven commented:
I'll look at this

ValenberghsSven commented:
Cypress is not picky, defining 2 unique calls can be done with plenty of wildcards.
```
cy.intercept('GET', `/publication-flows**?filter**case**${publicationNumber2}**`).as('getFilteredIdCase');
cy.intercept('GET', `/publication-flows**?filter**agenda*item*treatment**${publicationNumber2}**`).as('getFilteredIdTreatment');
```
and then chain awaited
```
cy.get(dependency.emberPowerSelect.searchInput).type(publicationNumber2)
.wait('@getFilteredIdCase')
.wait('@getFilteredIdTreatment');
```

The main issue was that we sometimes found the loading option, but we can ignore that option since it has a different class then a normal selectable option (the 2 `cy.wait` were not always enough..)
`cy.get(dependency.emberPowerSelect.optionLoadingMessage).should('not.exist');`

Part of the tests failing was also that we can click the `edit` button on the case panel while the `initFields` action is still running, and we are hitting a `null.localIdentifier` in `openEditingPanel()`.
So an additional fix is to either wait longer in cypress (checking the url changing wasn't enough), so opted to alter the code a bit, so we wait for the `initFields` before showing the edit button.

ValenberghsSven committed:
KAS-3419 fix test: await data loading before showing edit button + check for search option now showing

mdv-professional commented on app/components/publications/publication-flow-selector.js:
```diff
@@ -16,23 +16,54 @@ export default class PublicationsPublicationFlowSelectorComponent extends Compon
+      include: 'identification',
+    });
+
+    const [publicationsByCase, publicationsNotViaCouncilOfMinisters] = yield all([
```
Wat is de reden dat nu EmberConcurrency.all wordt gebruikt i.p.v. Promise.all?

erikap replied:
No specific reason since the passed promises aren't ember-concurrency tasks. Replaced with Promise.all().

ValenberghsSven commented on app/components/publications/batch-documents-publication-modal.js:
```diff
@@ -104,15 +106,52 @@ export default class PublicationsBatchDocumentsPublicationModalComponent extends
+          this.intl.t('unable-to-relink-publication-flow'),
+          this.intl.t('warning-title')
+        );
+        throw Error('Unable to relink publication flow');
```
In order to test his, I managed to hit his with concurrency, both browsers already searched the same number and tried to link 2 docs from different cases to the same publicationflow and the second one failed so it works.
Not sure why we are throwing an error though, nothing really catches this error so it just ends up as an `uncaught exception`.
Could just return to exit the method?

erikap replied:
Nice catch. It was throwing an error because I didn't expect to hit this case. Updated the inline documentation with the concurrency scenario and replaced the error-throwing with a regular return to exit the method :+1:

ValenberghsSven commented on app/services/publication-service.js:
```diff
@@ -190,9 +190,10 @@ export default class PublicationService extends Service {
-    const _case = await publicationFlow.case;
-    const subcases = await _case.subcases;
-    return !!subcases.length;
+    const agendaitem = await this.store.queryOne('agendaitem', {
+      'filter[treatments][publication-flows][:id:]': publicationFlow.id,
+    });
+    return agendaitem != null;
```
Oddly enough, this doesn't work.
the `queryOne` yields null even though I just linked a publication so it should return something.
`query` does return results

I think we are hitting a cache issue here:
`queryOne` only worked after I rebooted cache and resource.
I'm guessing the path to cache of agendaitem/treatments wasn't invalidated in this process so we keep hitting the stale one. Or something like that, kinda difficult to pinpoint
Using curl and adding something to the call also works to get results while the current call gets null

edit: I saw that there was no change to the "source", so newly linked publications still showed 'niet via MR'

ValenberghsSven replied:
"I'm guessing the path to cache of agendaitem/treatments wasn't invalidated in this process so we keep hitting the stale one"

This is kinda the same issue we have with the yggdrasil government cache?

ValenberghsSven replied:
I suggest using the same query as you do in the `linkPublicationFlow`
```
const agendaItemTreatment = await publicationFlow.agendaItemTreatment;
const agendaitem = await this.store.queryOne('agendaitem', {
  'filter[treatments][:id:]': agendaItemTreatment.id
});
```

erikap replied:
Changed the query to the one you proposed and documented the reasoning inline :+1: 

Sounds a lot like the Yggdrasil government cache issue. May be related to the fuzzy relation-cardinality between agendaitem and agenda-item-treatment (defined as `has-one` in mu-cl-resources, but `has-many` in data)?|
I hope the issue will  be "automatically" resolved by the upcoming data migration stories where we're going to split agenda-item-treatment and decision-activities which will result in a simple one-to-one relation between agendaitem and agenda-item-treatment. I would at least pause further investigation on the cache issue until that data migration is done.

ValenberghsSven replied:
"simple one-to-one relation between agendaitem and agenda-item-treatment"
this will not be the case since multiple agenda versions have different agendaitems that are linked to the same treatment, so it will still be a many-to-one. But since there will only ever be 1 treatment and multiple decisions, we wont have to deal with a many-to-many anymore

ValenberghsSven committed:
Merge branch 'development' of github.com:kanselarij-vlaanderen/frontend-kaleidos into feature/KAS-3419-link-publication-to-document-afterwards

erikap committed:
Merge branch 'development' into feature/KAS-3419-link-publication-to-document-afterwards

erikap committed:
Replace ember-concurrency all with Promise.all in publication-flow-selector

Since promises passed to all() are not cancellable there is no reason
to prefer ember-concurrency's all over regular Promise.all

erikap committed:
Document error in concurrency scenario when linking document to publication

erikap committed:
Use alternate query to get agendaitem for a publication-flow

erikap committed:
Merge branch 'feature/KAS-3419-link-publication-to-document-afterwards' of github.com:kanselarij-vlaanderen/frontend-kaleidos into feature/KAS-3419-link-publication-to-document-afterwards

ValenberghsSven commented:
jenkins ok buiten 1 gekende flaky deze keer, ik dacht dat die opgelost was..
